registries/adapters/apiv2_adapter.go

```diff
@@ -0,0 +1,220 @@
+const apiV2CatalogURL = "%v/v2/_catalog"
+
+// NewApiV2Adapter - creates and returns a *ApiV2Adapter ready to use.
+func NewApiV2Adapter(config Configuration) *ApiV2Adapter {
+	return &ApiV2Adapter{
+		Config: config,
+		client: oauth.NewClient(config.User, config.Pass, config.SkipVerifyTLS, config.URL),
+	}
+}
+
+// ApiV2Adapter - API V2 Adapter
+type ApiV2Adapter struct {
+	Config Configuration
+	client *oauth.Client
+}
+
+// ApiV2CatalogResponse - Catalog Response
+type ApiV2CatalogResponse struct {
+	Repositories []string `json:"repositories"`
+}
+
+// RegistryName - Retrieve the registry name
+func (r ApiV2Adapter) RegistryName() string {
+	if r.Config.URL.Host == "" {
+		return r.Config.URL.Path
+	}
+	return r.Config.URL.Host
+}
+
+// GetImageNames - retrieve the images
+func (r ApiV2Adapter) GetImageNames() ([]string, error) {
+	log.Debugf("%s - GetImageNames", r.Config.AdapterName)
+	log.Debugf("BundleSpecLabel: %s", BundleSpecLabel)
+
+	// Authorization
+	err := r.client.Getv2()
+	if err != nil {
+		log.Errorf("Failed to GET /v2 at %s - %v", apiV2CatalogURL, err)
+		return nil, err
+	}
+
+	// check if we're configured for specific images
+	if r.Config.Images != nil {
+		log.Debugf("Configured to use images: %v", r.Config.Images)
+		return r.Config.Images, nil
+	}
+
+	// discover images from URL
+	imageNames, err := r.discoverImages(fmt.Sprintf(apiV2CatalogURL, r.Config.URL))
+	if err != nil {
+		log.Error("Image Discovery Failed!")
+		return nil, err
+	}
+	return imageNames, nil
+}
+
+// discoverImages - Get Imagenames from the /v2/_catalog URL
+func (r ApiV2Adapter) discoverImages(catUrl string) ([]string, error) {
+	log.Debugf("%s - discoverImages", r.Config.AdapterName)
+
+	// Initial URL
+	if len(catUrl) == 0 {
+		return nil, errors.New("URL is empty!")
+	}
+	url := catUrl
+
+	// Get all Image Names until the 'Link' value is empty
+	// https://docs.docker.com/registry/spec/api/#pagination
+	var apbImages []string
+	for {
+		log.Debugf("discovering images from - %s", url)
+		images, linkStr, err := r.getNextImages(url)
+		if err != nil {
+			return nil, err
+		}
+
+		apbImages = append(apbImages, images.Repositories...)
+
+		// no more to get..
+		if len(linkStr) == 0 {
+			break
+		}
+
+		url = r.getNextImageUrl(linkStr)
+		if len(url) == 0 {
+			return nil, errors.New("Invalid Next Image URL")
+		}
+	}
+
+	return apbImages, nil
+}
+
+// getNextImages - Get the next 'Link' URL.
+func (r ApiV2Adapter) getNextImages(url string) (*ApiV2CatalogResponse, string, error) {
+	req, err := r.client.NewRequest(url)
+	if err != nil {
+		return nil, "", err
+	}
+
+	resp, err := r.client.Do(req)
+	if err != nil {
+		return nil, "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		log.Errorf("Failed to fetch catalog response from '%s'. Expected a 200 status and got: %v", url, resp.Status)
+		return nil, "", errors.New(resp.Status)
+	}
+
+	imageResp, err := ioutil.ReadAll(resp.Body)
```

Comment:
This is a nit because this is perfectly fine.

There are helper methods to do this
```go
imageList := ApiV2CatalogResponse{}
err = json.NewDecoder(resp.Body).Decode(&imageList)
```